File: osm/addr-osmify-php/addr-osmify.php
#!/usr/bin/php
<?php declare(strict_types=1);

/*
 * Copyright 2020 Miklos Vajna. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 *
 * Takes an OSM way ID and turns it into a string that is readable and
 * e.g. OsmAnd can parse it as well.
 */

// Send query to overpass turbo.
function query_turbo(string $query): string
{
    $url = 'http://overpass-api.de/api/interpreter';

    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $query);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

    return curl_exec($ch);
}

// Send query to nominatim.
function query_nominatim(string $query): string
{
    $url = 'https://nominatim.openstreetmap.org/search.php?';
    $params = [
        'q' => $query,
        'format' => 'json',
    ];
    $url .= http_build_query($params);

    return file_get_contents($url);
}

// Turn query into a coordinate + address string.
function osmify(string $query): void
{
    ini_set('user_agent', 'Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0');

    // Use nominatim to get the coordinates and the osm type/id.
    $elements = json_decode(query_nominatim($query));

    if (!count($elements)) {
        echo "No results from nominatim\n";

        return;
    }

    if (count($elements) > 1) {
        $buildings = [];

        foreach ($elements as $i) {
            if (!property_exists($i, 'class') || $i->class !== 'building') {
                break;
            }

            array_push($buildings, $i);
        }

        if (count($buildings)) {
            $elements = $buildings;
        }
    }

    $element = $elements[0];
    $lat = $element->lat;
    $lon = $element->lon;
    $object_type = $element->osm_type;
    $object_id = $element->osm_id;

    // Use overpass to get the properties of the object.
    $overpass_query = "[out:json];\n";
    $overpass_query .= '(';
    $overpass_query .= $object_type . '(' . $object_id . ');';
    $overpass_query .= ');';
    $overpass_query .= 'out body;';
    $j = json_decode(query_turbo($overpass_query));
    $elements = $j->elements;

    if (!count($elements)) {
        echo "No results from overpass\n";

        return;
    }

    $element = $elements[0];
    $city = $element->tags->{'addr:city'};
    $housenumber = $element->tags->{'addr:housenumber'};
    $postcode = $element->tags->{'addr:postcode'};
    $street = $element->tags->{'addr:street'};
    $addr = $postcode . ' ' . $city . ', ' . $street . ' ' . $housenumber;

    // Print the result.
    echo 'geo:' . $lat . ',' . $lon . ' (' . $addr . ")\n";
}

// Commandline interface to this module.
function main(): void
{
    if (count($_SERVER['argv']) > 1) {
        osmify($_SERVER['argv'][1]);
    } else {
        echo "usage: addr-osmify <query>\n";
        echo "\n";
        echo "e.g. addr-osmify 'Mészáros utca 58/a, Budapest'\n";
    }
}

main();

// vim:set shiftwidth=4 softtabstop=4 expandtab:
?>

File: osm/addr-osmify-py/addr_osmify.py
#!/usr/bin/env python3
#
# Copyright 2019 Miklos Vajna. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

"""
Takes an OSM way ID and turns it into a string that is readable and
e.g. OsmAnd can parse it as well.
"""

import json
import sys
import threading
import urllib.parse
import urllib.request
from typing import Dict
from typing import cast


def query_turbo(query: str) -> str:
    """Send query to overpass turbo."""
    url = "http://overpass-api.de/api/interpreter"

    sock = urllib.request.urlopen(url, bytes(query, "utf-8"))
    buf = sock.read()
    sock.close()

    return cast(str, buf.decode("utf-8"))


def query_nominatim(query: str) -> str:
    """Send query to nominatim."""
    url = "http://nominatim.openstreetmap.org/search.php?"
    params = {
        "q": query,
        "format": "json"
    }
    url += urllib.parse.urlencode(params)

    sock = urllib.request.urlopen(url)
    buf = sock.read()
    sock.close()

    return cast(str, buf.decode("utf-8"))


def osmify(query: str) -> str:
    """Turn query into a coordinate + address string."""
    # Use nominatim to get the coordinates and the osm type/id.
    elements = json.loads(query_nominatim(query))
    if not elements:
        return "No results from nominatim"

    if len(elements) > 1:
        # There are multiple elements, prefer buildings if possible.
        # Example where this is useful: 'Karinthy Frigyes út 18, Budapest'.
        buildings = [i for i in elements if "class" in i.keys() and i["class"] == "building"]
        if buildings:
            elements = buildings

    element = elements[0]
    lat = element["lat"]
    lon = element["lon"]
    object_type = element["osm_type"]
    object_id = element["osm_id"]

    # Use overpass to get the properties of the object.
    overpass_query = """[out:json];
(
    %s(%s);
);
out body;""" % (object_type, object_id)
    j = json.loads(query_turbo(overpass_query))
    elements = j["elements"]
    if not elements:
        return "No results from overpass"

    element = elements[0]
    city = element['tags']['addr:city']
    housenumber = element['tags']['addr:housenumber']
    postcode = element['tags']['addr:postcode']
    street = element['tags']['addr:street']
    addr = "%s %s, %s %s" % (postcode, city, street, housenumber)

    # Print the result.
    return "%s,%s (%s)" % (lat, lon, addr)


def worker(context: Dict[str, str]) -> None:
    """Wrapper around osmify() that has no return value."""
    context["out"] = osmify(context["in"])


def spinner(context: Dict[str, str], thread: threading.Thread) -> None:
    """Shows a spinner while osmify() is in progress."""
    spin_characters = "\\|/-"
    spin_index = 0
    while True:
        thread.join(timeout=0.1)
        if thread.is_alive():
            sys.stderr.write("\r [%s] " % spin_characters[spin_index])
            sys.stderr.flush()
            spin_index = (spin_index + 1) % len(spin_characters)
            continue

        sys.stderr.write("\r")
        sys.stderr.flush()
        print(context["out"])
        break


def main() -> None:
    """Commandline interface to this module."""
    if len(sys.argv) > 1:
        context = {"in": sys.argv[1]}
        thread = threading.Thread(target=worker, args=(context,))
        thread.start()
        spinner(context, thread)
    else:
        print("usage: addr-osmify <query>")
        print()
        print("e.g. addr-osmify 'Mészáros utca 58/a, Budapest'")


if __name__ == "__main__":
    main()

# vim:set shiftwidth=4 softtabstop=4 expandtab:
